### DISCUSSION TITLE
Support schima migration

### DISCUSSION THREAD

**Discussion by Mihai-B** [isar/isar#345] [Help]
Hi,

Really love the database and this repo does seem to be what I currently need, however I do have one question:

I've noticed from the documentation that Isar can support auto migration when a table is updated, but is there anyway we can do
this update manually? Something like when opening a connection to the database, when the system detects that the schema has changed then a function should be called where the "upgrade code" is there, and manual queries can be run to upgrade the tables.

**Comment by simc** [comment]
Hey @Mihai-B,
It is not possible to access the table with its old schema because the generated Dart code represents the new collection schema. Isar will handle all the index/link/property additions/deletions internally.

I think you want to migrate the actual data and not the schema. I didn't add data migration intentionally because it's quite easy to do it yourself and I didn't want to limit users to a predefined way of doing it. Just store the collection versions in a `"versions"` collection (or something similar) and invoke your migration logic on app start if the version is outdated.

**Comment by Mihai-B** [comment]
Hey @leisim , thnx for the reply.
 To be honest, my only concern is that over time I could end up leaving some "junk" in the repository, something like unused tables or columns.
The thing I am a bit concerned are this lines in the documentation:
```
Deleted fields will still remain in the database. You are not allowed to recreate deleted fields with a different type
```
and
```
Changing the type of fields in existing collections (even previously deleted ones)
```
So this sound to me like some "stuff" remains there in DB and it is not fully deleted. To do a bit of cleanup would go a long way, or even some option to just delete the table entirely and recreate it and have everything erased in that table.
But other than that, I really like this database.

**Reply by simc** [reply]
Yes, in order to avoid slow migration, Isar has a binary format that allows backwards compatible changes. The downside is that, like you said, data of removed field remains. Many databases work like this.

If you feel like your collection is too old and uses too much space after many schema changes, just copy your data to a fresh collection and clear the old one 🙂

**Reply by Mihai-B** [reply]
That is also an option. thank you for the input.

**Comment by Mihai-B** [comment]
One more thing @leisim , this is not related to my initial question:
I noticed that for Mac OS there is a binary that you can download and run the DB Inspector to inspect your database.
I am a linux user and while I can build and run the app(did not yet test it to see how I can view the data), there is not a quick app binary that I can just run. Is it ok with you if I create an app binary and release it on https://flathub.org ? I think it would help developers that just want to prototype something quickly.
I've never made a flatpak release before, so I don't know 100% sure how complicate it is(or possible for this particular scenario) but I would like to try it with your app if you give me permission

**Reply by simc** [reply]
That would be awesome! Best case would be if we could integrate it into the Isar CI to update the releases automatically.

There is some work here: https://github.com/isar/isar/pull/224
Maybe you can finish the PR? Just open a new one with your solution :+1:

**Reply by Mihai-B** [reply]
Ok. I will look into it
